### DISCUSSION TITLE
Virtual Machine - PoC

### DISCUSSION THREAD

**Discussion by mgrossmann** [adesutherland/CREXX#13] [Ideas]
Hello all,

At the moment I am busy reading Adrian's work. 
Slowly I think I understand how all this is supposed to work. :) I also realize that this is far beyond my abilities. 
But that has never stopped me from trying. :)

But now to the actual topic: 
@adesutherland, do you already have ideas, what concerns the bytecode? 
So do you already have instructions / opcodes in mind?

As I wrote yesterday, I would like to approach it iteratively. 
I would like to create an example file that contains the bytecode for, say (?)SAY. 
Then one could write a little module that showcases some aspects of the virtual machine. 

Mike

**Comment by adesutherland** [comment]
Hi Mike

I do have a list of opcodes, and I do understand what you are asking for.

Like baking bread we need a starter; should I hack together a starter VM which covers only a handful in opcodes and then we can incrementally improve and grow it together? Does that work? 

The good thing is then we can work on the parser/compiler, the rexx assembler (in rexx), and the VM in parallel - and this will be faster as more people can get involved.  Agree?

Adrian

**Reply by mgrossmann** [reply]
Hello Adrian,

that sounds great.👌

Mike

**Comment by mgrossmann** [comment]
Hi Adrian,

can you please give me more informations about the opcode for ILOAD, IADD and ISAY?
- How long are the opcodes? 4byte? (8byte long would be way to big, but would make it easy 
to map function pointer to it)
- The operands must be 8byte long? 
- what are the opcodes for the instructions above?

Peter and I are just fiddling around with this stuff.

Mike

**Reply by adesutherland** [reply]
Hello Mike & Peter!

First - Please send my apologies to your partners ... IT is better than a mistress :-)

You have exactly got the issue!

For a threaded interpreter in a 64bit world each opcode needs to be 64 bit! There is no point in compressing it because we need to replace opcodes (thread it) with 64bit addresses. This is also a issue for 64bit vs 32bit native code.

So the issue is that we might only have 100 or 200 opcodes - this is the balance we need to consider.

5 or 10 years ago I would have been convinced that a threaded interpreter would be much faster because of the way pipelining  works on a cpu - however very very cleaver people work on CPUs - today might be different!

I really want to run tests between a very simple classic bytecode interpreter where we can compress and minimise space (which also must have performance improvements) compared with a threaded approach - larger programs but should be faster.

My advice is to do the classic interpreter first (switch / case / goto) because this would be easy to convert to a threader ..

As for actual opcodes - I will make a list - but given that we only want a handful at the moment and it will be easy to change then just choose something arbitrary ...

Does that make sense?

Adrian

**Reply by rvjansen** [reply]
So what do we do for other architectures/bitnesses? For example, in 24-bit 370 code we don’t even have the whole address, but only the offset from the base segment.

**Reply by mgrossmann** [reply]
@rvjansen At the moment it is just about the byte code and how it is stored. If we, for example, using 64bit for an opcode, we are using only the right most byte, in the end. We can then also map this quite comfortably under VM/MVS to smaller addresses. Or have I misunderstood your question?

**Reply by mgrossmann** [reply]
@adesutherland ok, then we are riding the same train. :)

**Reply by mgrossmann** [reply]
@adesutherland I would like to use Golang for my Virtual Machine experiments. Are you fine with this? So we could easily cross compile for Linux, Windows and Mac.

**Reply by adesutherland** [reply]
When we save the binary to disk - 64bit might make for big files (but on modern machines ...). So we can consider a format to make the binaries smaller. Remember however the Java guys just ZIPed it up! We can do the same ...

**Reply by adesutherland** [reply]
Go for it with Golang - but we will need to port to C eventually

**Reply by mgrossmann** [reply]
If we ever will release a production grade Virtual Machine  implementation, we'll have to figure out how to do it anyway. If we want to provide a reference implementation for VM,MVS,LINUX,MAC&WIN, then Golang would already cover the whole open world. We can also do everything in C, but then we will not get around some #IFDEFs. :) 

I will make my first steps in Go for now, but I am open for everything. We should discuss that at a later time.

Regards
Mike

**Comment by mgrossmann** [comment]
Hey @adesutherland 

what are your thoughts about strings larger than 24 characters?
Should the string stored in the string_extend_buffer, then? 

MIkeG

**Comment by adesutherland** [comment]
Yes.

When I defined the structure I was envisaging a chain of them to store long strings, however I am thinking this is too complex.

I propose that we use the buffer for long strings. So a string is either stored in struct value (<24 characters) OR is stored in a single string_extend_buffer. The string_extend_buffer should be sized as a multiple of 256 (as a starting guess) - this to to make life easier/faster for malloc (in fact we will review the sizes of all mallocs in the vm to optimise them to avoid wrong sized gaps if that makes sense).

This also means the string_extend_buffer needs to store the buffer size. The good news is string growth can fill the buffer before we have to realloc.

Anyway - that's my thoughts.

Adrian

**Reply by mgrossmann** [reply]
> So a string is either stored in struct value (<24 characters)

Isn't it <=24 ?

> This also means the string_extend_buffer needs to store the buffer size.

Do we need another structure to achieve this? Or should we extend value with an string_num_extends field?


MIkeG
